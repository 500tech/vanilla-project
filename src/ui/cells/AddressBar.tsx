import React, { useState, useEffect } from "react";
import styled from "styled-components";
import { useLocation, useHistory } from "react-router-dom";
import { Input, Button } from "ui/atoms";

const Form = styled.form`
  display: flex;
`;

type Location = ReturnType<typeof useLocation>;

const locationToUrl = ({ pathname, search, hash }: Location) =>
  pathname + search + hash;

export function AddressBar() {
  const location = useLocation();
  const history = useHistory();
  const [url, setUrl] = useState(locationToUrl(location));
  useEffect(() => {
    setUrl(locationToUrl(location));
  }, [location]);
  return (
    <Form
      onSubmit={e => {
        e.preventDefault();
        history.push(url);
      }}
    >
      <Button type="button" onClick={() => history.goBack()}>
        Back
      </Button>
      <Input required value={url} onChange={e => setUrl(e.target.value)} />
      <Button type="button" onClick={() => history.goForward()}>
        Forward
      </Button>
    </Form>
  );
}
