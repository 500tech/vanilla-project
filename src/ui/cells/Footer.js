import React from 'react';
import { FooterSection } from 'ui/atoms';

export function Footer({ copyrightExpiary, name }) {
  const currentDate = new Date();
  const year = currentDate.getUTCFullYear();
  const showFooter = year <= copyrightExpiary;
  if (!showFooter) {
    return null;
  }
  return (
    <FooterSection>
      <small>
        &copy; Copyright {copyrightExpiary}, {name}
      </small>
    </FooterSection>
  );
}
