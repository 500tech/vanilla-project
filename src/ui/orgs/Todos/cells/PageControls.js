import React from 'react';
import { noop } from 'utils';
import { Button } from 'ui/atoms';

export function PageControls({
  onMarkAllAsDone = noop,
  onDeleteDone = noop,
  todos,
}) {
  return (
    <section className="controls">
      <Button
        onClick={() => onMarkAllAsDone()}
        disabled={todos.every(todo => todo.completed)}
      >
        Mark all as done
      </Button>
      <Button
        onClick={() => onDeleteDone()}
        disabled={!todos.some(todo => todo.completed)}
      >
        Delete all done
      </Button>
    </section>
  );
}
