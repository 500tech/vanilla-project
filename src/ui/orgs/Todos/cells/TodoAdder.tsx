import React, {
  useState,
  useEffect,
  useCallback,
  FunctionComponent
} from "react";
import { noop } from "utils";
import { Button, Input } from "ui/atoms";
import { useAutofocus } from "hooks/autofocus";

const isValid = (text: string) => text.length > 0;

interface Props {
  onAddTodo?: (text: string) => void;
}

export const TodoAdder: FunctionComponent<Props> = function TodoAdder({
  onAddTodo = noop
}) {
  const [text, setText] = useState("");
  const input = useAutofocus();
  useEffect(() => {
    if (text.match(/clear/i)) {
      setText("");
    }
  }, [text]);
  const submitText = useCallback(
    text => {
      if (isValid(text)) {
        onAddTodo(text);
        setText("");
      }
    },
    [onAddTodo]
  );
  useEffect(() => {
    const timer = setTimeout(submitText, 3000, text);
    return () => clearTimeout(timer);
  }, [submitText, text]);

  return (
    <section>
      <form
        onSubmit={e => {
          e.preventDefault();
          submitText(text);
        }}
      >
        <Input
          ref={input}
          type="text"
          name="todo"
          placeholder="Write up your todos"
          required
          onChange={e => setText(e.target.value)}
          value={text}
        />
        <Button disabled={!isValid(text)}>Add</Button>
      </form>
    </section>
  );
};
