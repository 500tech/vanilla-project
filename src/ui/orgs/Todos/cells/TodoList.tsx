import React, { FunctionComponent } from "react";
import { TodoItem } from "ui/orgs/Todos/mols/TodoItem";
import { Todo } from "types";

interface Props {
  todos: Todo[];
  onToggle?: (todoId: Todo["id"]) => void;
  onDelete?: (todoId: Todo["id"]) => void;
}

export const TodoList: FunctionComponent<Props> = function TodoList({
  todos,
  onToggle,
  onDelete
}) {
  return (
    <ul>
      {todos.map(todo => (
        <TodoItem
          key={todo.id}
          todo={todo}
          onToggle={onToggle}
          onDelete={onDelete}
        />
      ))}
    </ul>
  );
};
