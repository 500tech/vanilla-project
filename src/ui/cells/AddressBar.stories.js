import React from "react";
import { MemoryRouter, useLocation } from "react-router-dom";
import { AddressBar } from "ui/cells/AddressBar";

const LocationLabel = () => {
  const location = useLocation();
  return (
    <p>
      <code>{JSON.stringify(location)}</code>
    </p>
  );
};

function withRouter(storyFn) {
  return (
    <MemoryRouter>
      <section>{storyFn()}</section>
      <LocationLabel />
    </MemoryRouter>
  );
}

export default {
  title: "cells/AddressBar",
  decorators: [withRouter]
};

export const Usage = () => <AddressBar />;
