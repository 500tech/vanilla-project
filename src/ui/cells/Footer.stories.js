import React from "react";
import { number, text } from "@storybook/addon-knobs";
import { Footer } from "ui/cells/Footer";

export default {
  title: "cells/Footer"
};

export const WithProps = () => (
  <Footer
    name={text("Name", "foobar co")}
    copyrightExpiary={number("Year", 2058, {
      range: true,
      min: 2018,
      max: 2060,
      step: 1
    })}
  />
);
