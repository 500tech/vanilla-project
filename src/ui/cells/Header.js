import React from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { HeaderSection } from 'ui/atoms';
import { AddressBar } from 'ui/cells/AddressBar';
import { useThemeService } from 'services/theme';

const Nav = styled.nav`
  &&& {
    a {
      margin-left: 5px;
      color: ${props => props.theme.colors.primary};
    }
  }
`;

export function Header({ children }) {
  const { toggleTheme, themeName, setTheme, themeNames } = useThemeService();
  return (
    <HeaderSection>
      <h1 onClick={toggleTheme}>{children}</h1>
      <select value={themeName} onChange={e => setTheme(e.target.value)}>
        {themeNames.map(name => (
          <option key={name} value={name}>
            {name}
          </option>
        ))}
      </select>
      <AddressBar />
      <Nav>
        <Link to="/">Home</Link>
        <Link to="/todos">Todos</Link>
      </Nav>
    </HeaderSection>
  );
}
