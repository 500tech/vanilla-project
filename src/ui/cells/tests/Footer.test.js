import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { mountWithTheme } from "test-utils";
import { Footer } from "ui/cells/Footer";

describe("<Footer>", () => {
  it("should render correctly", () => {
    const footer = <Footer copyrightExpiary={2099} name="foobar" />;
    const renderedTree = shallow(footer);
    expect(toJSON(renderedTree)).toMatchSnapshot();
  });

  it("Should show the correct year and name", () => {
    const wrapper = mountWithTheme(
      <Footer copyrightExpiary={2099} name="foobar" />
    );

    const footer = wrapper.getDOMNode();
    expect(footer).toHaveTextContent(/foobar/);
    expect(footer).toHaveTextContent(/2099/);
  });

  it("Should only appear if copyright is valid", () => {
    const wrapper = mountWithTheme(
      <Footer copyrightExpiary={1999} name="foobar" />
    );
    const footer = wrapper.getDOMNode();
    expect(footer).toBeNull();
  });
});
