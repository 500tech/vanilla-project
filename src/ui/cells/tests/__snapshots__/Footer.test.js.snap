// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer> should render correctly 1`] = `
<styled.footer>
  <small>
    Â© Copyright 
    2099
    , 
    foobar
  </small>
</styled.footer>
`;
